python/social_bot/tasks.py

```diff
@@ -851,3 +872,119 @@ def task_specific_observation(self, agent):
+        agent_sentence = yield
+        obj_pos = self._random_move_object(
+            target=self._object,
+            random_range=self._object_random_range,
+            center_pos=np.array([0, 0]),
+            mini_distance=self._mini_distance,
+            height=0)
+        goal_pos = self._random_move_object(
+            target=self._goal,
+            random_range=self._place_to_random_range,
+            center_pos=obj_pos[:2],
+            mini_distance=self._mini_distance,
+            height=0)
+        steps = 0
+        while steps < self._max_steps:
+            steps += 1
+            # get positions
+            obj_pos, _ = self._object.get_pose()
+            finger_l_pos, _ = self._agent.get_link_pose(self._finger_link_l)
+            finger_r_pos, _ = self._agent.get_link_pose(self._finger_link_r)
+            finger_pos = (
+                np.array(finger_l_pos) + np.array(finger_r_pos)) / 2.0
+            # get contacts
+            l_contact = self._agent.get_contacts(
+                'finger_cnta_l', 'wood_cube_5cm::link::collision')
+            r_contact = self._agent.get_contacts(
+                'finger_cnta_r', 'wood_cube_5cm::link::collision')
+            # check distance and contacts
+            dist = np.linalg.norm(np.array(obj_pos) - goal_pos)
+            palm_dist = np.linalg.norm(
+                np.array(obj_pos) - np.array(finger_pos))
+            gripping_feature_num = 1.0 * (
+                obj_pos[2] > 0.02) + 1.0 * l_contact + 1.0 * r_contact
+            # yield rewards
+            if gripping_feature_num >= 3 and dist < self._success_distance_thresh:
+                logging.debug("object has been successfuly placed")
+                reward = 100.0 if self._reward_shaping else 1.0
+                agent_sentence = yield TeacherAction(
+                    reward=reward, sentence="well done", done=True)
+            else:
+                reward = (gripping_feature_num - dist -
+                          palm_dist) if self._reward_shaping else 0
+                agent_sentence = yield TeacherAction(reward=reward, done=False)
+        yield TeacherAction(reward=-1.0, sentence="failed", done=True)
+
+    def task_specific_observation(self, agent):
+        """
+        Args:
+            agent (GazeboAgent): the agent
+        Returns:
+            np.array, the extra observations will be added into the observation
+        """
+        # Use 3 position of the links to uniquely determine the 6 + 2 DoF gripper
+        finger_l_pos, _ = agent.get_link_pose(self._finger_link_l)
+        finger_r_pos, _ = agent.get_link_pose(self._finger_link_r)
+        palm_pos, _ = agent.get_link_pose(self._palm_link)
+        # goal pos and object pos
+        goal_pos, _ = self._goal.get_pose()
+        obj_pos, obj_rot = self._object.get_pose()
+        # contacts
+        finger_contacts = np.array([
+            agent.get_contacts('finger_cnta_l', 'wood_cube_5cm::link::collision'),
```

Comment:
Is it possible to get "wood_cube_5cm" from self._object?  This would make the change easier to extend to picking up other objects.